src/main/java/org/apache/cassandra/distributed/api/LogAction.java

```diff
@@ -0,0 +1,352 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.api;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public interface LogAction
+{
+    /**
+     * @return current position of the iterator
+     * @see LogAction#grep(long, String)
+     * @see LogAction#watchFor(long, String)
+     */
+    long mark();
+
+    LineIterator matching(long startPosition, Predicate<String> fn);
+
+    default LineIterator matching(Predicate<String> fn){
+        return matching(Internal.DEFAULT_START_POSITION, fn);
+    }
+
+    default List<String> watchFor(long startPosition, Duration timeout, Predicate<String> fn) throws TimeoutException
+    {
+        long nowNanos = System.nanoTime();
+        long timeoutAtNanos = nowNanos + timeout.toNanos();
+        long previousPosition = startPosition;
+        List<String> matches = new ArrayList<>();
+        while (System.nanoTime() <= timeoutAtNanos)
+        {
+            if (previousPosition == mark())
+            {
+                // still matching... wait a bit
+                Internal.sleepUninterruptibly(1, TimeUnit.SECONDS);
+                continue;
+            }
+            // position not matching, try to read
+            try (LineIterator it = matching(previousPosition, fn))
+            {
+                while (it.hasNext())
+                    matches.add(it.next());
+                if (!matches.isEmpty())
+                    return matches;
+                previousPosition = it.mark();
+            }
+        }
+        throw new TimeoutException();
+    }
+
+    default List<String> watchFor(Duration timeout, Predicate<String> fn) throws TimeoutException {
```

Comment:
in writing tests here, I find that its hard to chain the file offset as we don't return it; curious if we should have a `LogResult` class which stores the offset and matches?

Reply:
Agree. We need to expose the previous position to the caller. Otherwise, it has no idea when the last position is and it is unable to seek to the last position.